test/test_core.jl

```diff
@@ -378,29 +382,41 @@ end
     l = @optic _.x
     @test l(t) === 1
 
+    @test set(t, @optic(getfield(_, :x)), :hello) === (x=:hello, y=2)
+
     # do we want this to throw an error?
     @test_throws ArgumentError (@set t.z = 3)
+    @test_throws Exception set(t, @optic(getfield(_, :z)), 3)
 end
 
 struct CustomProperties
     _a
     _b
+    _c
 end
 
-function ConstructionBase.setproperties(o::CustomProperties, patch::NamedTuple)
-    CustomProperties(get(patch, :a, getfield(o, :_a)),
-                     get(patch, :b, getfield(o, :_b)))
+Base.propertynames(o::CustomProperties) = (:a, :b)
 
-end
+Base.getproperty(o::CustomProperties, k) = getfield(k == :a ? :_a : :_b)
 
-ConstructionBase.constructorof(::Type{CustomProperties}) = error()
+ConstructionBase.setproperties(o::CustomProperties, patch::NamedTuple) = 
+    CustomProperties(get(patch, :a, getfield(o, :_a)),
+                     get(patch, :b, getfield(o, :_b)),
+                     getfield(o, :_c))
 
 @testset "setproperties overloading" begin
-    o = CustomProperties("A", "B")
+    o = CustomProperties("A", "B", "C")
     o2 = @set o.a = :A
-    @test o2 == CustomProperties(:A, "B")
+    @test o2 == CustomProperties(:A, "B", "C")
     o3 = @set o.b = :B
-    @test o3 == CustomProperties("A", :B)
+    @test o3 == CustomProperties("A", :B, "C")
+
```

Comment:
Some inference tests would be good.

Reply:
Somethink like
```julia
function set_field_a(obj, val)
    @set getfield(obj,:a) = val
end
@inferred set_field_a(obj, val)
```

Reply:
That is a good point. `getfield(_, :a)` will store the `:a` as a runtime symbol. Which is bad for "long range" inference like here. 
Maybe a better approach is to introduce `FieldLens` analogous to `PropertyLens`. It can store the field as part of the type and enable long range inference.

I think manipulating fields is a rare enough use case to justify ugly syntax like `@set FieldLens(:a)(obj) = val`.

Reply:
I think the analogy with `IndexLens` is flawed. First of all `getindex` is usually type stable (think of Vectors, Dicts etc.), while `getfield` and `getproperty` are rarely type stable.
So it is a sane default for `IndexLens` to store the index as a field and a sane default for `PropertyLens` to store it in the type. If the default is not ideal, say because you want to manipulate Tuples over long range, you can request that:
```julia
using StaticNumbers
lens = @optic _[static(2)] # Tuple access will be type stable with this
```
So I think the analogy between field access  and property access is much closer than between field access and index access. Therefore we should mimic PropertyLens not IndexLens here.